core/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java

```diff
@@ -1032,7 +1063,15 @@ public MetaData build() {
                 }
             }
             aliasAndIndexLookup = Collections.unmodifiableSortedMap(aliasAndIndexLookup);
-            return new MetaData(clusterUUID, version, transientSettings, persistentSettings, indices.build(), templates.build(), customs.build(), allIndices, allOpenIndices, allClosedIndices, aliasAndIndexLookup);
+            if (graveyardBuilder != null) {
+                // if the index graveyard has been modified, add the new version to the custom map
+                customs.put(IndexGraveyard.TYPE, graveyardBuilder.build());
+            } else if (customs.get(IndexGraveyard.TYPE) == null) {
+                // it was never part of the customs map, so add a fresh entry
+                customs.put(IndexGraveyard.TYPE, IndexGraveyard.builder().build());
+            }
```

Comment:
This adds complexity here in the `MetaData.Builder`, in order to make it easier to work with `IndexGraveyard` in `MetaDataDeleteIndexService` and the various tests.  I could go either way on this.